In [1]:
import os
import sys

sys.path.append(os.path.abspath(".."))
from src.data_processing import *
from src.feature_engineering import *
from src.utils.helpers import *
from src.models.xgboost import train_xgboost_model, predict_xgboost
from src.models.lightgbm import train_lightgbm, predict_lightgbm
from src.models.random_forest import train_random_forest, predict_random_forest

In [2]:
X_train, X_val, y_train, y_val, X_test = prepare_data(categorical_handling = 'object_to_category')


# Model Selection

In [3]:
xgboost_model = train_xgboost_model(X_train, X_val, y_train, y_val)

[0]	train-auc:0.76684	eval-auc:0.76329
[50]	train-auc:0.90242	eval-auc:0.89945
[100]	train-auc:0.91681	eval-auc:0.91285
[150]	train-auc:0.92664	eval-auc:0.92103
[200]	train-auc:0.93298	eval-auc:0.92594
[250]	train-auc:0.93882	eval-auc:0.93037
[300]	train-auc:0.94316	eval-auc:0.93369
[350]	train-auc:0.94732	eval-auc:0.93672
[400]	train-auc:0.95042	eval-auc:0.93873
[450]	train-auc:0.95399	eval-auc:0.94125
[499]	train-auc:0.95696	eval-auc:0.94317
AUC: 0.9432
Accuracy: 0.9805
Precision: 0.9415
Recall: 0.4859
F1 Score: 0.6410
Model xgboost_model saved to /Users/elnur/Desktop/nexi_project/ieee-fraud-detection/src/models/saved_models.


In [5]:
lightgbm_model = train_lightgbm(X_train, X_val, y_train, y_val)

Training until validation scores don't improve for 50 rounds
[50]	training's auc: 0.901235	valid_1's auc: 0.899802
[100]	training's auc: 0.922768	valid_1's auc: 0.918413
[150]	training's auc: 0.934838	valid_1's auc: 0.928499
[200]	training's auc: 0.944204	valid_1's auc: 0.934596
[250]	training's auc: 0.950167	valid_1's auc: 0.93788
[300]	training's auc: 0.954746	valid_1's auc: 0.940581
[350]	training's auc: 0.959693	valid_1's auc: 0.943441
[400]	training's auc: 0.962416	valid_1's auc: 0.944906
[450]	training's auc: 0.964983	valid_1's auc: 0.946233
[500]	training's auc: 0.968211	valid_1's auc: 0.948892
Did not meet early stopping. Best iteration is:
[500]	training's auc: 0.968211	valid_1's auc: 0.948892
AUC: 0.9489
Accuracy: 0.9813
Precision: 0.9381
Recall: 0.5141
F1 Score: 0.6642
Model lightgbm_model saved to /Users/elnur/Desktop/nexi_project/ieee-fraud-detection/src/models/saved_models.


In [4]:
lightgbm_model_with_categorical_encoding = train_lightgbm(X_train, X_val, y_train, y_val)

Training until validation scores don't improve for 50 rounds
[50]	training's auc: 0.899498	valid_1's auc: 0.898615
[100]	training's auc: 0.922159	valid_1's auc: 0.917503
[150]	training's auc: 0.93402	valid_1's auc: 0.927446
[200]	training's auc: 0.941726	valid_1's auc: 0.932754
[250]	training's auc: 0.946653	valid_1's auc: 0.936363
[300]	training's auc: 0.951807	valid_1's auc: 0.939549
[350]	training's auc: 0.955891	valid_1's auc: 0.942409
[400]	training's auc: 0.95911	valid_1's auc: 0.944266
[450]	training's auc: 0.962257	valid_1's auc: 0.946004
[500]	training's auc: 0.965378	valid_1's auc: 0.947337
Did not meet early stopping. Best iteration is:
[500]	training's auc: 0.965378	valid_1's auc: 0.947337
AUC: 0.9473
Accuracy: 0.9804
Precision: 0.9331
Recall: 0.4896
F1 Score: 0.6422
Model lightgbm_model saved to /Users/elnur/Desktop/nexi_project/ieee-fraud-detection/src/models/saved_models.


In [3]:
rf_model = train_random_forest(X_train, X_val, y_train, y_val)

AUC: 0.8900
Accuracy: 0.8796
Precision: 0.1916
Recall: 0.7308
F1 Score: 0.3037
Model random_forest_model saved to /Users/elnur/Desktop/nexi_project/ieee-fraud-detection/src/models/saved_models.


# Model Evaluation and Comparison

We implemented several models for fraud detection and compared their performance using key metrics. Here’s a summary of our results:

**XGBoost (with Object-to-Category Encoding)**
- **AUC:** 0.9432  
- **Accuracy:** 0.9805  
- **Precision:** 0.9415  
- **Recall:** 0.4859  
- **F1 Score:** 0.6410  

**LightGBM (with Object-to-Category Encoding)**
- **AUC:** 0.9489  
- **Accuracy:** 0.9813  
- **Precision:** 0.9381  
- **Recall:** 0.5141  
- **F1 Score:** 0.6642  

**LightGBM (with Label Encoding)**
- **AUC:** 0.9473  
- **Accuracy:** 0.9804  
- **Precision:** 0.9331  
- **Recall:** 0.4896  
- **F1 Score:** 0.6422  

**Random Forest (with Label Encoding)**
- **AUC:** 0.8900  
- **Accuracy:** 0.8796  
- **Precision:** 0.1916  
- **Recall:** 0.7308  
- **F1 Score:** 0.3037  

---

**Analysis & Insights**

- **Boosting Models (XGBoost & LightGBM):**  
  Both XGBoost and LightGBM using object-to-category encoding perform very well (AUC > 0.94). LightGBM slightly outperforms XGBoost in AUC and F1 score and achieves a marginally higher recall. This suggests that preserving the original categorical structure is beneficial for this dataset.

- **Encoding Impact:**  
  For LightGBM, using object-to-category encoding yields better results than label encoding, underscoring the importance of retaining detailed categorical distinctions.  
  In contrast, Random Forest, which required label encoding, performs notably worse, indicating that boosting methods are more adept at capturing the complex relationships in the data.

- **Trade-Offs:**  
  Although precision is high across boosting models, recall remains moderate (around 48–51%). In fraud detection, this trade-off is common since fraudulent cases are rare. Balancing false negatives (missed fraud) and false positives is critical, and the F1 score reflects a reasonable compromise.


# Trying test set

In [3]:
#load the model
xgboost_model = load_model('xgboost_model')

Model xgboost_model loaded from /Users/elnur/Desktop/nexi_project/ieee-fraud-detection/src/models/saved_models.


In [4]:
xgboost_y_pred = predict_xgboost(xgboost_model, X_test)

In [7]:
submission = prepare_my_submission(xgboost_y_pred, 'xgboost_model')

In [8]:
submission.head()

TransactionID   isFraud
0        3663549  0.006961
1        3663550  0.005903
2        3663551  0.005462
3        3663552  0.002001
4        3663553  0.004546

# Final Submission 

- Model used: **XGBoost**
  - **Public Score**: 0.8732
  - **Private Score**: 0.8072

- Thresholded version (at 0.5) yielded:
  - **Public Score**: 0.7283
  - **Private Score**: 0.6633

**Direct probability output outperforms thresholded predictions for AUC-based evaluation.**